avocado/plugins/spawners/remote.py

```diff
@@ -0,0 +1,210 @@
+            runtime_task.spawner_handle = slot
+            return await fn(self, runtime_task)
+
+    return wrapper
+
+
+class RemoteSpawner(Spawner, SpawnerMixin):
+
+    description = "Remote (host) based spawner"
+    METHODS = [SpawnMethod.STANDALONE_EXECUTABLE]
+    slots_cache = {}
+
+    @staticmethod
+    async def run_remote_cmd_async(session, command, timeout):
+        loop = asyncio.get_event_loop()
+        try:
+            status, output = await loop.run_in_executor(
+                None, session.cmd_status_output, command, timeout
+            )
+        except exceptions.ShellTimeoutError:
+            status, output = 2, f"Remote command timeout of {timeout} reached"
+        except exceptions.ShellProcessTerminatedError:
+            status, output = 2, f"Remote command terminated prematurely"
+        return status, output
+
+    @contextlib.contextmanager
+    def reserve_slot(self, runtime_task):
+        """
+        Reserve a free or custom remote host slot for the runtime task.
+
+        :param runtime_task: runtime task to reserve the slot for
+        :type runtime_task: :py:class:`avocado.core.task.runtime.RuntimeTask`
+        :yields: a free slot to use if such was found
+        :raises: :py:class:`RuntimeError` if no free slot could be found
+
+        This will either use a runtime cache to find a free remote host slot to
+        run the task in or use a custom hostname/slot ID to allow for custom
+        schedulers to make their own decisions on which hosts to run and when.
+        """
+        if len(RemoteSpawner.slots_cache) == 0:
+            # TODO: consider whether to provide persistence across runs via external storage
+            for session_slot in self.config.get("spawner.remote.slots"):
+                if not session_slot:
+                    continue
+                with open(session_slot, "r") as f:
+                    session_data = json.load(f)
+                session = remote.remote_login(**session_data)
+                RemoteSpawner.slots_cache[session] = False
+
+        if runtime_task.spawner_handle is not None:
+            slot = runtime_task.spawner_handle
+        else:
+            slots = RemoteSpawner.slots_cache
+            for key, value in slots.items():
+                if not value:
+                    slot = key
+                    slots[key] = True
+                    break
+            else:
+                raise RuntimeError(
+                    "No free slot available for the task, are "
+                    "you running with more processes than slots?"
+                )
+
+        try:
+            yield slot
+        finally:
+            RemoteSpawner.slots_cache[slot] = False
+
+    @staticmethod
+    def is_task_alive(runtime_task):
+        if runtime_task.spawner_handle is None:
+            return False
+
+        session = runtime_task.spawner_handle
+
+        status, _ = session.cmd_status_output("pgrep -r R,S -f task-run")
```

Comment:
Can't we match the task's identifier `runtime_task.task.identifier` to the one given on the command line (as in `avocado-runner-$kind task-run -i $identification`)?

Reply:
I guess we can, I am just mimicking some previous implementations of the podman spawner from a few years ago that likely have changed in this direction. Will take a look too.

Reply:
Ok done, resolving.